File: admin-ui/src/types/policy.ts
import type { DecisionFunctionSummary } from './decisionFunction'

export type PolicyType = 'row_filter' | 'column_mask' | 'column_allow' | 'column_deny' | 'table_deny'

export type AssignmentScope = 'all' | 'user' | 'role'

export interface TargetEntry {
  schemas: string[]
  tables: string[]
  columns?: string[]
}

export interface PolicyAssignmentResponse {
  id: string
  policy_id: string
  policy_name: string
  data_source_id: string
  datasource_name: string
  user_id: string | null
  username: string | null
  role_id: string | null
  role_name: string | null
  assignment_scope: AssignmentScope
  priority: number
  created_at: string
}

export interface PolicyResponse {
  id: string
  name: string
  description: string | null
  policy_type: string
  targets: TargetEntry[]
  definition: Record<string, string> | null
  is_enabled: boolean
  version: number
  decision_function_id?: string | null
  decision_function?: DecisionFunctionSummary | null
  assignment_count: number
  created_by: string
  updated_by: string
  created_at: string
  updated_at: string
  assignments?: PolicyAssignmentResponse[]
}

export interface CreatePolicyPayload {
  name: string
  description?: string
  policy_type: PolicyType
  is_enabled: boolean
  targets: TargetEntry[]
  definition?: Record<string, string> | null
  decision_function_id?: string | null
}

export interface UpdatePolicyPayload {
  name?: string
  description?: string
  policy_type?: PolicyType
  is_enabled?: boolean
  targets?: TargetEntry[]
  definition?: Record<string, string> | null
  decision_function_id?: string | null
  version: number
}

export interface AssignPolicyPayload {
  policy_id: string
  user_id?: string | null
  role_id?: string | null
  scope?: AssignmentScope
  priority: number
}

// ---------- Anchor coverage (edit-time silent-deny warning) ----------

/// Per-(table, column) verdict for whether a row-filter policy's column
/// reference will resolve at query time. Tagged-union JSON keyed by `kind`.
export type AnchorCoverageVerdict =
  | { kind: 'on_table'; column: string }
  | {
      kind: 'anchor_walk'
      column: string
      via_relationship_id: string
      via_child_column: string
      via_parent_column: string
      parent_schema: string
      parent_table: string
    }
  | { kind: 'anchor_alias'; column: string; actual_column_name: string }
  | { kind: 'missing_anchor'; column: string }
  | {
      kind: 'missing_column_on_alias_target'
      column: string
      actual_column_name: string
    }

export interface AnchorCoverageTableEntry {
  data_source_id: string
  data_source_name: string
  schema: string
  table: string
  verdicts: AnchorCoverageVerdict[]
}

export interface PolicyAnchorCoverageResponse {
  policy_id: string
  policy_type: string
  coverage: AnchorCoverageTableEntry[]
}

File: proxy/src/entity/policy.rs
use sea_orm::entity::prelude::*;
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
#[sea_orm(table_name = "policy")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: Uuid,
    #[sea_orm(unique)]
    pub name: String,
    pub description: Option<String>,
    /// "row_filter" | "column_mask" | "column_allow" | "column_deny" | "table_deny"
    pub policy_type: String,
    /// JSON array of TargetEntry (schemas, tables, columns?)
    pub targets: String,
    /// JSON definition (filter_expression or mask_expression). Null for non-expression types.
    pub definition: Option<String>,
    pub is_enabled: bool,
    pub version: i32,
    pub decision_function_id: Option<Uuid>,
    pub created_by: Uuid,
    pub updated_by: Uuid,
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(has_many = "super::policy_version::Entity")]
    PolicyVersion,
    #[sea_orm(has_many = "super::policy_assignment::Entity")]
    PolicyAssignment,
    #[sea_orm(
        belongs_to = "super::decision_function::Entity",
        from = "Column::DecisionFunctionId",
        to = "super::decision_function::Column::Id"
    )]
    DecisionFunction,
}

impl Related<super::policy_version::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::PolicyVersion.def()
    }
}

impl Related<super::policy_assignment::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::PolicyAssignment.def()
    }
}

impl Related<super::decision_function::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::DecisionFunction.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
